undercrawler/middleware/autologin.py

```diff
@@ -55,44 +59,83 @@ def __init__(self, autologin_url, auth_cookies=None, logout_url=None,
         else:
             self.auth_cookies = None
             self.logged_in = False
-        self.logout_url = logout_url
 
     @classmethod
     def from_crawler(cls, crawler):
         if not crawler.settings.getbool('AUTOLOGIN_ENABLED'):
             raise NotConfigured
-        return cls(**{name.lower(): crawler.settings.get(name) for name in [
-            'AUTOLOGIN_URL', 'AUTH_COOKIES', 'LOGOUT_URL', 'SPLASH_URL',
-            'LOGIN_URL', 'USERNAME', 'PASSWORD', 'USER_AGENT',
-            'AUTOLOGIN_DOWNLOAD_DELAY']})
+        return cls(crawler.settings['AUTOLOGIN_URL'], crawler)
 
     def process_request(self, request, spider):
         ''' Login if we are not logged in yet.
         '''
-        if '_autologin' in request.meta:
+        if '_autologin' in request.meta or request.meta.get('skip_autologin'):
             return
-        # Save original request to be able to retry it in case of logout
-        req_copy = request.replace(meta=deepcopy(request.meta))
-        req_copy.callback = req_copy.errback = None
-        request.meta['_autologin'] = autologin_meta = {'request': req_copy}
-
-        if not self.logged_in:
-            self.auth_cookies = self.get_auth_cookies(request.url)
-            self.logged_in = True
-        elif self.logout_url and self.logout_url in request.url:
-            logger.debug('Ignoring logout request %s', request.url)
-            raise IgnoreRequest
-
-        # Always apply self.cookies, so that changes in self.auth_cookies
-        # are applied to all future requests immediately.
-        if self.auth_cookies:
-            request.cookies = self.auth_cookies
-            autologin_meta['cookie_dict'] = {
-                c['name']: c['value'] for c in self.auth_cookies}
-
-    def get_auth_cookies(self, url):
+        if self.skipped:
+            request.meta['autologin_active'] = False
+            return
+        elif self.logged_in:
+            request.meta['autologin_active'] = True
+            if self.logout_url and self.logout_url in request.url:
+                logger.debug('Ignoring logout request %s', request.url)
+                raise IgnoreRequest
+            # Save original request to be able to retry it in case of logout
+            req_copy = request.replace(meta=deepcopy(request.meta))
+            req_copy.callback = req_copy.errback = None
+            request.meta['_autologin'] = autologin_meta = {'request': req_copy}
+            # TODO - it should be possible to put auth cookies into the
+            # cookiejar in process_response (but also check non-splash)
+            if self.auth_cookies:
+                request.cookies = self.auth_cookies
+                autologin_meta['cookie_dict'] = {
+                    c['name']: c['value'] for c in self.auth_cookies}
+        else:
+            self._enqueue(request)
+            if self.waiting_for_login:
+                raise IgnoreRequest
```

Comment:
It tool a while to get used to such style, but at the end it's better, I think. The only gotcha is that at one point the traceback was incorrect, is it worth a bug report?